### DISCUSSION TITLE
Burmese script marking character placement problem with Harfbuzz / Speedata

### DISCUSSION THREAD

**Discussion by Nainoia-Inc-Admin** [harfbuzz/harfbuzz#4784] [Q&A] [answered]
I am using Speedata / Hardbuzz to produce a PDF of the Sanskrit New Testament in Burmese text using the Padauk font from Google. The word below in found in the middle of Matthew 4:21 and the marking characters are placed properly with the Padauk font in the browser. However, Harfbuzz does not place the marking characters properly. What can I do to help you solve the problem. See below.

သာရ္ဒ္ဓံ

![speedata-harfbuzz-burmese-script-problem](https://github.com/harfbuzz/harfbuzz/assets/46004014/da65fa13-b564-4104-b878-13faf2a17302)

**Comment by khaledhosny** [comment]
If I download Padauk from Google Fonts, I don’t get the 2nd output with `hb-view`, which suggests it is a Speeddta issue not a HarfBuzz one. I get the 3rd output with `hb-view`, but I get also the same output when using CoreText and DirectWrite, which suggests it is a font issue if this is indeed the incorrect output.

Interestingly, if I download it from SIL’s website, I always get the 2nd output unless I tell `hb-view` to not use the `graphite2` shaper, then I get the 3rd output. The font from Google Fonts does not have Graphite table.

I suggest reporting this to the font developers, either this is a font bug or bad input, but ot does not look like there is a HarfBuzz issue here.

**Reply by Nainoia-Inc-Admin** [reply]
Do you know the correct placement of the marking characters? I was thinking that the first rendering was correct because the marking characters do not overlap, the browser display them that way and also the Sanskrit website places them that way at http://www.sanskritbible.in/assets/txt/burmese/40004.html.

If the browser know how to place the marking characters correctly doesn't that suggest Harfbuzz is the problem and not the font?

**Reply by khaledhosny** [reply]
It is likely your browser is also using HarfBuzz (both Chrome and Firefox use HarfBuzz) but a different font.

**Comment by alerque** [comment]
I can confirm basically what Khaled pointed out, the Graphite tables don't give the same results as the OpenType ones.

Here is a SILE document testing both ways:

```sile
\begin{document}
\language[main=my]
\font[family=Padauk,size=72pt]

\set[parameter=harfbuzz.subshapers,value=ot]
သာရ္ဒ္ဓံ

\set[parameter=harfbuzz.subshapers,value=graphite2]
သာရ္ဒ္ဓံ

\end{document}
```

![image](https://github.com/harfbuzz/harfbuzz/assets/173595/1ed866c0-dfde-4aa0-afed-dd42b46d52ec)

**Comment by alerque** [comment]
Also a note, you said it works fine in "the browser", but in my Browser (Firefox) even your GitHub post renders like your second example:

![image](https://github.com/harfbuzz/harfbuzz/assets/173595/56c77d9b-23ea-4c3d-bcba-9bd5050dede2)

Inspecting the DOM in Firefox this seems to be rendering with *Noto Sans Myanmar*. I'm not sure if this is another broken font or if perhaps your input string is really the trouble.

**Reply by Nainoia-Inc-Admin** [reply]
Also note that Padauk5000 from SIL does NOT work properly, but Padauk5001 does work properly and also Google Padauk works properly in the browser. So what version of Padauk are you using?

**Reply by alerque** [reply]
```console
$ sile burmese.sil -d versions                                                                                                                      ❌USAGE  108ms ✦1
SILE v0.15.4 (LuaJIT 2.1.1700206165) [Rust]
<burmese.sil> as sil
[1]
Harfbuzz version: 9.0.0
Shapers enabled: graphite2, ot, fallback
ICU support enabled

Fonts used:
/usr/share/fonts/TTF/Padauk-Regular.ttf:0       Version 5.000
```

I'll go bump the Padauk package on Arch Linux and report back again.

**Reply by alerque** [reply]
I just updated my Padauk to 5.001 and got exactly the same rendering both for OpenType and Graphite shaping [from my comment above](https://github.com/harfbuzz/harfbuzz/discussions/4784#discussioncomment-9930105) that was rendered with 5.000.

**Reply by Nainoia-Inc-Admin** [reply]
Hmmm. I explicitly loaded the Padauk5001 font in my HTML and FF rendered it properly in my tests, though I settled on the Google Padauk which also displayed properly, but Speedata/Harfbuzz did not with either font. What does Google Padauk do for you? 

https://fonts.google.com/specimen/Padauk

Though I just visited the Google Padauk font page and tested သာရ္ဒ္ဓံ and it displayed improperly, the 3rd option in the original post above. So now I am totally confused.

Here is my webpage that displays the font properly for me, Matthew 4:21 middle word...
(edit) > also of note that my website uses a woff2 generated from Font Squirrel with the Google Padauk TTF is input.

https://stage.aionianbible.org/Bibles/Sanskrit---Burmese-Script/Matthew/4/parallel-English---Aionian-Bible

And here is the source website which displays the word properly

http://www.sanskritbible.in/assets/txt/burmese/40004.html

**Reply by khaledhosny** [reply, marked as answer]
> Here is my webpage that displays the font properly for me, Matthew 4:21 middle word...

The Padauk WOFF2 file used in that page does not have any Burmese characters, so your browser is using a fallback font. You can use Firefox web Developer Tools to find out the actual font(s) used in the page: https://devtoolstips.org/tips/en/list-used-fonts/

**Reply by Nainoia-Inc-Admin** [reply]
Yes I see that now. So that also points to the answer that the problem is a bad font file. Does anyone have a recommended font for Burmese? How can I get my hands on the Firefox fallback font called "Myanmar Text"? Maybe this is it https://font.download/font/myanmar-text, but I need a Myanmar font with open licensing. So perhaps a better font would solve the problem with both Speedata and HTML.

**Reply by devosb** [reply]
Maybe a bad font file on your website, as documented by @khaledhosny , but the [upstream Padauk font from SIL](https://software.sil.org/padauk/) (which I now maintain) should work better, at least with OpenType shaping, and includes `ttf`, `woff`, and `woff2` files. I would be curious if you use the WOFF2 file from SIL how your website looks.

I don't know of a better, open font to use. Here is a sample of various fonts, Padauk has two columns, one for Graphite (GR) shaping, one for OpenType (OT) shaping.
![image](https://github.com/harfbuzz/harfbuzz/assets/12450072/5b8aea6c-3ae4-47ed-892e-cba142f69bff)
This matches what @khaledhosny and @alerque found, that the Graphite shaping in Padauk has a problem. Firefox, while it does use HarfBuzz for OpenType shaping, also includes Graphite. Chrome does not include Graphite.

Myanmar Text does display as you desire (Windows 10, in both Word 365 and Notepad)
![image](https://github.com/harfbuzz/harfbuzz/assets/12450072/e536ae73-4a82-4155-9890-9689d217d939)

So I think all of this explains the difference in shaping that we are observing (which font and what shaping rules (OpenType or Graphite)) are being used.

I would not have called Myanmar Text a Firefox fallback font. This [font is from Microsoft](https://learn.microsoft.com/en-us/typography/font-list/myanmar-text), and comes with Windows when you enable Burmese language support. I suspect the website mentioned with this font is not a legitimate source for the font. I am not surprised that Firefox, on Windows, uses this font as a fallback for Burmese script text. But Firefox on other platforms (like Linux, and I suspect macOS) are not going to have this font.

So the OpenType shaping of Padauk might not be wrong, but as you mention, the shaping from Myanmar Text is preferred. Do you know if this is unique to the Sanskrit language (in Burmese script), or would other languages in Burmese script also desire the form as seen in Myanmar Text?

In my testing the shaping does not seem to depend if the word is in the middle or end of the line. The text uses spaces, which surprises me for this script. @Nainoia-Inc-Admin can you clarify if you get different shaping depending on where the word is in the text and/or confirm that spaces are desired?

**Reply by Nainoia-Inc-Admin** [reply]
I think I answered your two question below, but answering here also for clarity and hopes that you can fix the SIL Padauk font. I have 4 different Myanmar/Burmese Bible texts I manage and this particular unicode sequence is found only in the Sanskrit NT Burmese script Bible text.  And yes the Padauk v5.0001 font performs differently depending where the word is on the line.  If the middle of the line, then your first example above. If at the end of the line then your second example above.  I tried to explain that in the original post above, but might not have been clear.  Hope that helps you work a solution.  And finally I can confirm that spaces are allowed in the text, at least my text sources seem to think so.

**Comment by pgundlach** [comment]
I am late in the game ... 

The speedata Publisher uses HarfBuzz, although I cannot guarantee that I use HarfBuzz correctly. I usually check with FontGoggles and nowadays with https://simoncozens.github.io/sploot/ and at least in this case get the same results in both as the PDF output (the last image in the first post in this thread).

**Comment by behdad** [comment]
> Does anyone have a recommended font for Burmese? 

You can use Noto Sans Myanmar:

https://fonts.google.com/noto/specimen/Noto+Sans+Myanmar?query=noto+sans+myanmar

**Reply by devosb** [reply]
Both Noto Sans and Serif Myanmar have clashes, as shown in [my comment](https://github.com/harfbuzz/harfbuzz/discussions/4784#discussioncomment-9959810). Did I misunderstand something?

Now that I re-read this discussion, I see that the form that Myanmar Text provides is not necessarily the preferred form, it is just that is what was seen in a browser, due to font fall back and the website not sending a working font.

While you could use the WOFF2 files from the Padauk release from SIL, they would have Graphite, so your website would display the clash when using Firefox. I would suggest using the Padauk font from Google Fonts, and use Google's CSS instructions to have your website load the font from Google.

**Comment by Nainoia-Inc-Admin** [comment]
Thanks for all the input and this seem to be the conclusion:

**I found and tested 4/5 Myanmar Burmese fonts:**
https://software.sil.org/padauk/ version 5.001  (this appears to be the same as https://fonts.google.com/specimen/Padauk)
https://fonts.google.com/noto/specimen/Noto+Sans+Myanmar?query=myanmar
https://fonts.google.com/noto/specimen/Noto+Serif+Myanmar?query=myanmar
Microsoft mmrtext.ttf and  mmrtextb.ttf  (tested though restrictive copyright prevents my usage)

**Results:**
All fonts provided all the glyphs needed in my 4 different Bible texts. 

The questioned unicode sequence of \x{1039}\x{1012}\x{1039}\x{1013}\x{1036} used in the word သာရ္ဒ္ဓံ was only found in one of my Bible texts, the Sanskrit NT Burmese script text. See middle word at Matthew 4:21, http://www.sanskritbible.in/assets/txt/burmese/40004.html

Only the Microsoft mmrtext.ttf properly displayed this unicode sequence without clashing in both Speedata/Harfbuzz and Firefox HTML. All the other fonts clashed the marking characters in both Speedata/Harfbuzz and HTML.

My preferred font for appearance is https://software.sil.org/padauk/ version 5.001 and since that font maintainer is in the discussion maybe a version 5.002 is a possibility to fix the above unicode character marker clashing?

**Reply by devosb** [reply]
Yes, Padauk on Google Fonts is the same font as what SIL releases, except that Google removed the Graphite tables from their version. They might have made other, much smaller, changes in the font metadata.

Maybe I am getting confused by what you mean by clashing. I would call the example of PadaukOT a _minor_ clash, as the U+1012 MYANMAR LETTER DA slightly overlaps the tail of U+101B MYANMAR LETTER RA. I would call the example of PadaukGR a _major_  clash since the U+1013 MYANMAR LETTER DHA completely over-strikes DA.

So when you say clashing do you mean _minor_ or _major_ clashing?

I would also try using Chrome and Edge web browsers. I would add to your test data the string `RenderingUnknown` this should change to to indicate if Graphite or OpenType shaping is being used.

**Reply by Nainoia-Inc-Admin** [reply]
They are both major clashes.

As you say the PadaukGR is a major clash since the U+1013 MYANMAR LETTER DHA completely over-strikes DA. However, in the case of the PadaukOT the U+1012 MYANMAR LETTER DA slightly overlaps the tail of U+101B MYANMAR LETTER RA. The reason this is major is because this marker is out of place and if the word happens to be at the end of a line Speedata/Harfbuzz will put the character into the margin. Speedata doesn't know the marker is in the margin where it doesn't belong. So if the application does not allow rogue characters in the margin the problem becomes major. And even if line breaks happened properly the marker is still improperly placed.

The improper placement in the margin is what alerted me to the problem in the first place.

**Reply by devosb** [reply]
We updated the Graphite code and the latest build of the font now has both the Graphite and OpenType shaping somewhat close. You can get the latest font from the [README.MD](https://github.com/silnrsi/font-padauk/blob/master/README.md) file in the font repo, click on the [button](https://build.palaso.org/viewType.html?buildTypeId=Fonts_Padauk&guest=1) to the right of the heading _Project status._ There should be TTF, WOFF, and WOFF2.

I realize the shaping is not ideal, and we are disscusioning what to do next and when.

**Reply by Nainoia-Inc-Admin** [reply]
Thanks for working on this. I see that the unicode marks no longer clobber each other, though they are not stacked properly yet either. I also see that there are other unicode marks effected or should be effected by this new rule as well. This might be a much bigger deal.

Here is a picture of the culprit word using Padauk50002 with marks not stacked and rolling into the margin.

![Padauk50002 sample](https://github.com/harfbuzz/harfbuzz/assets/46004014/1f4731dc-79eb-499c-999d-bf485db42cdd)

Here it is properly displayed with mmrtext.ttf

![mmrtext sample](https://github.com/harfbuzz/harfbuzz/assets/46004014/5b2b9825-57fb-44a5-b4ec-12f9b9703ffc)

I am also noticing other errors in words in Matthew 4 with problems.
Another word from Matthew 4 with Padauk50002

![Padauk50002 other bad](https://github.com/harfbuzz/harfbuzz/assets/46004014/2758ceaa-a11a-4658-b100-b90b32aca9b7)

Yet here is it displayed in mmrtext.ttf

![mmrtext other good](https://github.com/harfbuzz/harfbuzz/assets/46004014/848a116c-fc7f-4598-940c-dc44c5d98ca6)

A summary of the problem is that when there are multiple marking characters the mmrtext.ttf font stacks them so they are all there with out clobbering each other. However, Padauk50001 allowed marks to clobber each other, and now Padauk50002 prevents clobbering, but does not stack the marks with the associated letter.

What info can I provide to help?

**Reply by alerque** [reply]
This discussion probably could move to an issue report on the Padauk repository. Having looked at the commit log  since the last release it might be time to nudge for a new release too, :wink: :wink:.

Meanwhile wouldn't the issue of "rolling into the margin" as you put it be a problem with the typesetting engine not taking the shape reported by Harfbuzz into account when justifying the line, not an issue with the font at all?

**Reply by Nainoia-Inc-Admin** [reply]
Can someone move the whole discussion or do I need to start a new issue at https://github.com/silnrsi/font-padauk?

The typesetting engine might need improvement, but the font also is not properly specifying how to properly place the marks in association with their letters. We know this because the Microsoft mmrtext.ttf font does work properly with the same typesetting engine as shown in my above examples.

**Reply by alerque** [reply]
Not all of this is directly relevant for Padauk and I don't think we have anyone that has enough permissions on both repos to migrate wholesale. I'd just open a new issue picking up the topic and link back to this for context for how we got here.

What is that screen shot of typeset text from? A browser or something else?

**Reply by pgundlach** [reply]
> Meanwhile wouldn't the issue of "rolling into the margin" as you put it be a problem with the typesetting engine not taking the shape reported by Harfbuzz into account when justifying the line, not an issue with the font at all?

The speedata Publisher takes the advance width into account, not any glyph widths. And the advance width reported by harfbuzz is 0. So I am not sure how this _should_ be handled.

**Reply by Nainoia-Inc-Admin** [reply]
Discussion continuing here https://github.com/silnrsi/font-padauk/issues/52

**Comment by mhosken** [comment]
Looks like a request for a contextual ligature for the subjoined consonants. The context being a single width base.

**Reply by Nainoia-Inc-Admin** [reply]
Is that how the Microsoft mmrtext.ttf font solves the problem? Because that font works properly.

**Comment by Nainoia-Inc-Admin** [comment]
Discussion continuing here https://github.com/silnrsi/font-padauk/issues/52
